src/crypto/crypto_hkdf.cc

```diff
@@ -103,20 +103,49 @@ bool HKDFTraits::DeriveBits(
   EVPKeyCtxPointer ctx =
       EVPKeyCtxPointer(EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr));
   if (!ctx || !EVP_PKEY_derive_init(ctx.get()) ||
-      !EVP_PKEY_CTX_hkdf_mode(ctx.get(),
-                              EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
       !EVP_PKEY_CTX_set_hkdf_md(ctx.get(), params.digest) ||
-      !EVP_PKEY_CTX_set1_hkdf_salt(
-          ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
-      !EVP_PKEY_CTX_set1_hkdf_key(
-          ctx.get(),
-          reinterpret_cast<const unsigned char*>(params.key->GetSymmetricKey()),
-          params.key->GetSymmetricKeySize()) ||
       !EVP_PKEY_CTX_add1_hkdf_info(
           ctx.get(), params.info.data<unsigned char>(), params.info.size())) {
     return false;
   }
 
+  if (params.key->GetSymmetricKeySize() != 0) {
+    if (!EVP_PKEY_CTX_hkdf_mode(ctx.get(),
+                                EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
+        !EVP_PKEY_CTX_set1_hkdf_salt(
+            ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
```

Comment:
Does a non-zero-length key work with a 0-length salt?

Reply:
Yes.

Reply:
https://datatracker.ietf.org/doc/html/rfc5869#section-2.2

Reply:
I know it's allowed, I am just wondering if OpenSSL implements it that way. The RFC's statement `if not provided` could be interpreted as "if the user does not call `EVP_PKEY_CTX_set1_hkdf_salt`".

Now I am confused as to whether an empty salt is the same as passing no salt in our current implementation.

Reply:
Our implementation does not allow to omit passing salt, we require the argument, albeit we allow it to be zero-length.

Reply:
> Does a non-zero-length key work with a 0-length salt?

that's also covered by a passing wpt btw